crates/agentgateway/src/http/csrf.rs

```diff
@@ -0,0 +1,158 @@
+						);
+					},
+				}
+			},
+			None => {
+				// No Sec-Fetch-Site header - fallthrough to Origin check
+			},
+		}
+
+		// 3. Get Origin header once
+		let origin = get_origin_header(req)?;
+
+		// 4. If no Origin header, allow
+		if origin.is_none() {
+			return Ok(Default::default());
+		}
+
+		let origin_str = origin.unwrap();
+
+		// 5. Check if Origin matches Host header
+		let target_origin = get_target_origin(req)?;
+		if origins_match(&origin_str, &target_origin) {
+			return Ok(Default::default());
+		}
+
+		// 6. Check trusted origins as last resort
+		if self.is_request_exempt(req)? {
+			return Ok(Default::default());
+		}
+
+		// Request failed all checks - reject
+		self.create_forbidden_response("Cross-origin request detected".to_string())
+	}
+
+	fn is_request_exempt(&self, req: &Request) -> Result<bool, filters::Error> {
+		if let Some(origin) = get_origin_header(req)? {
+			for allowed_origin in &self.additional_origins {
+				if origin == *allowed_origin {
+					return Ok(true);
+				}
+			}
+		}
+		Ok(false)
+	}
+
+	/// Create a 403 Forbidden response
+	fn create_forbidden_response(&self, message: String) -> Result<PolicyResponse, filters::Error> {
+		let response = ::http::Response::builder()
+			.status(StatusCode::FORBIDDEN)
+			.body(crate::http::Body::from(message))?;
+		Ok(PolicyResponse {
+			direct_response: Some(response),
+			response_headers: None,
+		})
+	}
+}
+
+/// Check if the HTTP method is a safe method
+fn is_safe_method(method: &Method) -> bool {
+	matches!(method, &Method::GET | &Method::HEAD | &Method::OPTIONS)
+}
+
+/// Extract the Origin header value
+fn get_origin_header(req: &Request) -> Result<Option<String>, filters::Error> {
+	if let Some(origin_value) = req.headers().get(header::ORIGIN) {
+		let origin_str = origin_value.to_str().map_err(|_| {
+			filters::Error::InvalidFilterConfiguration("Invalid Origin header".to_string())
+		})?;
+
+		// Handle "null" origin as no origin
+		if origin_str == "null" {
+			return Ok(None);
+		}
+
+		return Ok(Some(origin_str.to_string()));
+	}
+	Ok(None)
+}
+
+/// Extract the Sec-Fetch-Site header value
+fn get_sec_fetch_site_header(req: &Request) -> Result<Option<String>, filters::Error> {
+	if let Some(sec_fetch_site_value) = req.headers().get("sec-fetch-site") {
+		let sec_fetch_site_str = sec_fetch_site_value.to_str().map_err(|_| {
+			filters::Error::InvalidFilterConfiguration("Invalid Sec-Fetch-Site header".to_string())
+		})?;
+		return Ok(Some(sec_fetch_site_str.to_string()));
+	}
+	Ok(None)
+}
+
+/// Extract the target origin from the request
+fn get_target_origin(req: &Request) -> Result<String, filters::Error> {
+	// After URI normalization, check authority first
+	if let Some(authority) = req.uri().authority() {
+		let scheme = req.uri().scheme_str().unwrap_or("http");
+		return Ok(format!("{}://{}", scheme, authority));
+	}
+
+	// Fallback to Host header
+	let host_value = req
```

Comment:
I don't think we need or want the fallback. We already do `normalize_uri` earlier in the proxy path which guarantees `authority()` is set to the proper value.